File: apps/web/src/shared/api/migrationBoard.ts
import type { components } from "@country-decision-atlas/contracts/generated/types";
import { csrfHeaders } from "../auth/session";
import { apiDelete, apiGet, apiPatch, apiPost, queryString } from "./http";

export type MigrationBoardPostListResponse =
  components["schemas"]["MigrationBoardPostListResponse"];
export type MigrationBoardPostDetail =
  components["schemas"]["MigrationBoardPostDetail"];
export type MyMigrationBoardPost =
  components["schemas"]["MyMigrationBoardPost"];
export type MyMigrationBoardPostListResponse =
  components["schemas"]["MyMigrationBoardPostListResponse"];
export type CreateMigrationBoardPostRequest =
  components["schemas"]["CreateMigrationBoardPostRequest"];
export type UpdateMigrationBoardPostRequest =
  components["schemas"]["UpdateMigrationBoardPostRequest"];
export type ContactRequestListResponse =
  components["schemas"]["ContactRequestListResponse"];
export type ContactRequestResponse =
  components["schemas"]["ContactRequestResponse"];
export type CreateContactRequestRequest =
  components["schemas"]["CreateContactRequestRequest"];
export type CreateMigrationBoardReportRequest =
  components["schemas"]["CreateMigrationBoardReportRequest"];
export type MigrationBoardReportResponse =
  components["schemas"]["MigrationBoardReportResponse"];
export type CompanionMatchesResponse =
  components["schemas"]["CompanionMatchesResponse"];
export type AdminMigrationBoardPostListResponse =
  components["schemas"]["AdminMigrationBoardPostListResponse"];
export type MigrationBoardReportListResponse =
  components["schemas"]["MigrationBoardReportListResponse"];
export type ModerateMigrationBoardPostRequest =
  components["schemas"]["ModerateMigrationBoardPostRequest"];
export type ReviewMigrationBoardReportRequest =
  components["schemas"]["ReviewMigrationBoardReportRequest"];
export type BlockedUserListResponse =
  components["schemas"]["BlockedUserListResponse"];

export type BoardPostFilters = {
  destination_country?: string;
  origin_country?: string;
  route_id?: string;
  scenario?: string;
  persona?: string;
  timeline_window?: string;
  migration_stage?: string;
  companion_goal?: string;
  household_type?: string;
  preferred_language?: string;
  tag?: string;
  visibility?: string;
  limit?: number;
  offset?: number;
};

export function listBoardPosts(
  filters: BoardPostFilters = {},
): Promise<MigrationBoardPostListResponse> {
  return apiGet<MigrationBoardPostListResponse>(
    `/api/v1/migration-board/posts${queryString(filters)}`,
    { headers: csrfHeaders() },
  );
}

export function getBoardPost(
  postId: string,
): Promise<MigrationBoardPostDetail> {
  return apiGet<MigrationBoardPostDetail>(
    `/api/v1/migration-board/posts/${postId}`,
    {
      headers: csrfHeaders(),
    },
  );
}

export function listMyBoardPosts(): Promise<MyMigrationBoardPostListResponse> {
  return apiGet<MyMigrationBoardPostListResponse>(
    "/api/v1/me/migration-board/posts",
    {
      headers: csrfHeaders(),
    },
  );
}

export function createBoardPost(
  payload: CreateMigrationBoardPostRequest,
): Promise<MyMigrationBoardPost> {
  return apiPost<MyMigrationBoardPost, CreateMigrationBoardPostRequest>(
    "/api/v1/me/migration-board/posts",
    payload,
    { headers: csrfHeaders() },
  );
}

export function updateBoardPost(
  postId: string,
  payload: UpdateMigrationBoardPostRequest,
): Promise<MyMigrationBoardPost> {
  return apiPatch<MyMigrationBoardPost, UpdateMigrationBoardPostRequest>(
    `/api/v1/me/migration-board/posts/${postId}`,
    payload,
    { headers: csrfHeaders() },
  );
}

export function submitBoardPost(
  postId: string,
): Promise<{ post: MyMigrationBoardPost }> {
  return apiPost<{ post: MyMigrationBoardPost }, Record<string, never>>(
    `/api/v1/me/migration-board/posts/${postId}/submit`,
    {},
    { headers: csrfHeaders() },
  );
}

export function archiveBoardPost(
  postId: string,
): Promise<{ post: MyMigrationBoardPost }> {
  return apiPost<{ post: MyMigrationBoardPost }, Record<string, never>>(
    `/api/v1/me/migration-board/posts/${postId}/archive`,
    {},
    { headers: csrfHeaders() },
  );
}

export function listCompanionMatches(
  postId?: string,
): Promise<CompanionMatchesResponse> {
  const path = postId
    ? `/api/v1/me/migration-board/posts/${postId}/matches`
    : "/api/v1/me/migration-board/matches";
  return apiGet<CompanionMatchesResponse>(path, { headers: csrfHeaders() });
}

export function createContactRequest(
  postId: string,
  payload: CreateContactRequestRequest,
): Promise<ContactRequestResponse> {
  return apiPost<ContactRequestResponse, CreateContactRequestRequest>(
    `/api/v1/migration-board/posts/${postId}/contact-requests`,
    payload,
    { headers: csrfHeaders() },
  );
}

export function listIncomingContactRequests(): Promise<ContactRequestListResponse> {
  return apiGet<ContactRequestListResponse>(
    "/api/v1/me/migration-board/contact-requests/incoming",
    { headers: csrfHeaders() },
  );
}

export function listOutgoingContactRequests(): Promise<ContactRequestListResponse> {
  return apiGet<ContactRequestListResponse>(
    "/api/v1/me/migration-board/contact-requests/outgoing",
    { headers: csrfHeaders() },
  );
}

export function acceptContactRequest(
  requestId: string,
): Promise<{ request: ContactRequestResponse }> {
  return apiPost<
    { request: ContactRequestResponse },
    { response_note?: string | null }
  >(
    `/api/v1/me/migration-board/contact-requests/${requestId}/accept`,
    {},
    { headers: csrfHeaders() },
  );
}

export function declineContactRequest(
  requestId: string,
): Promise<{ request: ContactRequestResponse }> {
  return apiPost<
    { request: ContactRequestResponse },
    { response_note?: string | null }
  >(
    `/api/v1/me/migration-board/contact-requests/${requestId}/decline`,
    {},
    { headers: csrfHeaders() },
  );
}

export function cancelContactRequest(
  requestId: string,
): Promise<{ request: ContactRequestResponse }> {
  return apiPost<{ request: ContactRequestResponse }, Record<string, never>>(
    `/api/v1/me/migration-board/contact-requests/${requestId}/cancel`,
    {},
    { headers: csrfHeaders() },
  );
}

export function reportPost(
  postId: string,
  payload: CreateMigrationBoardReportRequest,
): Promise<MigrationBoardReportResponse> {
  return apiPost<
    MigrationBoardReportResponse,
    CreateMigrationBoardReportRequest
  >(`/api/v1/migration-board/posts/${postId}/report`, payload, {
    headers: csrfHeaders(),
  });
}

export function blockUser(userId: string, reason?: string): Promise<unknown> {
  return apiPost<unknown, { reason?: string }>(
    `/api/v1/me/migration-board/blocks/${userId}`,
    { reason },
    { headers: csrfHeaders() },
  );
}

export function unblockUser(userId: string): Promise<void> {
  return apiDelete<void>(`/api/v1/me/migration-board/blocks/${userId}`, {
    headers: csrfHeaders(),
  });
}

export function listBlockedUsers(): Promise<BlockedUserListResponse> {
  return apiGet<BlockedUserListResponse>("/api/v1/me/migration-board/blocks", {
    headers: csrfHeaders(),
  });
}

export function listAdminBoardPosts(
  status?: string,
): Promise<AdminMigrationBoardPostListResponse> {
  return apiGet<AdminMigrationBoardPostListResponse>(
    `/api/v1/admin/migration-board/posts${queryString({ status })}`,
    { headers: csrfHeaders() },
  );
}

export function approveAdminBoardPost(postId: string): Promise<unknown> {
  return apiPost<unknown, Record<string, never>>(
    `/api/v1/admin/migration-board/posts/${postId}/approve`,
    {},
    { headers: csrfHeaders() },
  );
}

export function rejectAdminBoardPost(
  postId: string,
  payload: ModerateMigrationBoardPostRequest,
): Promise<unknown> {
  return apiPost<unknown, ModerateMigrationBoardPostRequest>(
    `/api/v1/admin/migration-board/posts/${postId}/reject`,
    payload,
    { headers: csrfHeaders() },
  );
}

export function hideAdminBoardPost(
  postId: string,
  payload: ModerateMigrationBoardPostRequest,
): Promise<unknown> {
  return apiPost<unknown, ModerateMigrationBoardPostRequest>(
    `/api/v1/admin/migration-board/posts/${postId}/hide`,
    payload,
    { headers: csrfHeaders() },
  );
}

export function listAdminBoardReports(): Promise<MigrationBoardReportListResponse> {
  return apiGet<MigrationBoardReportListResponse>(
    "/api/v1/admin/migration-board/reports",
    { headers: csrfHeaders() },
  );
}

export function resolveAdminBoardReport(
  reportId: string,
  payload: ReviewMigrationBoardReportRequest,
): Promise<unknown> {
  return apiPost<unknown, ReviewMigrationBoardReportRequest>(
    `/api/v1/admin/migration-board/reports/${reportId}/resolve`,
    payload,
    { headers: csrfHeaders() },
  );
}

export function dismissAdminBoardReport(
  reportId: string,
  payload: ReviewMigrationBoardReportRequest,
): Promise<unknown> {
  return apiPost<unknown, ReviewMigrationBoardReportRequest>(
    `/api/v1/admin/migration-board/reports/${reportId}/dismiss`,
    payload,
    { headers: csrfHeaders() },
  );
}

export const migrationBoardApi = {
  listBoardPosts,
  getBoardPost,
  listMyBoardPosts,
  createBoardPost,
  updateBoardPost,
  submitBoardPost,
  archiveBoardPost,
  listCompanionMatches,
  createContactRequest,
  listIncomingContactRequests,
  listOutgoingContactRequests,
  acceptContactRequest,
  declineContactRequest,
  cancelContactRequest,
  reportPost,
  blockUser,
  unblockUser,
  listBlockedUsers,
  listAdminBoardPosts,
  approveAdminBoardPost,
  rejectAdminBoardPost,
  hideAdminBoardPost,
  listAdminBoardReports,
  resolveAdminBoardReport,
  dismissAdminBoardReport,
};

File: apps/api/app/api/v1/migration_board.py
from app.core.auth import CurrentUser, bearer_scheme
from app.core.database import get_connection
from app.core.rbac import require_user
from app.schemas.migration_board import (
    ArchiveMigrationBoardPostResponse,
    BlockedUserListResponse,
    BlockedUserResponse,
    BlockUserRequest,
    CompanionMatchesResponse,
    ContactRequestActionRequest,
    ContactRequestActionResponse,
    ContactRequestListResponse,
    ContactRequestResponse,
    CreateContactRequestRequest,
    CreateMigrationBoardPostRequest,
    CreateMigrationBoardReportRequest,
    MigrationBoardPostDetail,
    MigrationBoardPostListResponse,
    MigrationBoardReportResponse,
    MyMigrationBoardPost,
    MyMigrationBoardPostListResponse,
    SubmitMigrationBoardPostResponse,
    UpdateMigrationBoardPostRequest,
)
from app.services import auth as auth_service, migration_board as service
from fastapi import APIRouter, Depends, Query, Security, status
from fastapi.security import HTTPAuthorizationCredentials
from psycopg import Connection
from typing import Annotated, Any


router = APIRouter(tags=["migration-board"])


def optional_current_user(
    connection: Annotated[Connection[Any], Depends(get_connection)],
    credentials: Annotated[
        HTTPAuthorizationCredentials | None, Security(bearer_scheme)
    ],
) -> CurrentUser | None:
    if credentials is None or not credentials.credentials:
        return None
    result = auth_service.validate_session_token(
        connection, credentials.credentials
    )
    user = result["user"]
    return CurrentUser(
        id=user["id"],
        email=user["email"],
        display_name=user["display_name"],
        role=user["role"],
        status=user["status"],
    )


@router.get(
    "/migration-board/posts", response_model=MigrationBoardPostListResponse
)
def list_board_posts(
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser | None, Depends(optional_current_user)],
    destination_country: str | None = Query(None),
    origin_country: str | None = Query(None),
    route_id: str | None = Query(None),
    scenario: str | None = Query(None),
    persona: str | None = Query(None),
    timeline_window: str | None = Query(None),
    migration_stage: str | None = Query(None),
    companion_goal: str | None = Query(None),
    household_type: str | None = Query(None),
    preferred_language: str | None = Query(None),
    tag: str | None = Query(None),
    visibility: str | None = Query(None),
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
) -> dict[str, Any]:
    return service.list_public_posts(
        connection,
        current_user=current_user,
        filters={
            "destination_country_slug": destination_country,
            "origin_country_slug": origin_country,
            "route_id": route_id,
            "scenario_slug": scenario,
            "persona_slug": persona,
            "timeline_window": timeline_window,
            "migration_stage": migration_stage,
            "companion_goal": companion_goal,
            "household_type": household_type,
            "preferred_language": preferred_language,
            "tag": tag,
            "visibility": visibility,
        },
        limit=limit,
        offset=offset,
    )


@router.get(
    "/migration-board/posts/{post_id}", response_model=MigrationBoardPostDetail
)
def get_board_post(
    post_id: str,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser | None, Depends(optional_current_user)],
) -> dict[str, Any]:
    return service.get_public_post(
        connection, post_id=post_id, current_user=current_user
    )


@router.get(
    "/me/migration-board/posts",
    response_model=MyMigrationBoardPostListResponse,
)
def list_my_board_posts(
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    return service.list_my_posts(connection, current_user)


@router.post(
    "/me/migration-board/posts",
    response_model=MyMigrationBoardPost,
    status_code=status.HTTP_201_CREATED,
)
def create_my_board_post(
    payload: CreateMigrationBoardPostRequest,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    post = service.create_user_post(
        connection, current_user=current_user, payload=payload
    )
    connection.commit()
    return post


@router.get(
    "/me/migration-board/posts/{post_id}",
    response_model=MyMigrationBoardPost,
)
def get_my_board_post(
    post_id: str,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    return service.get_my_post(
        connection, current_user=current_user, post_id=post_id
    )


@router.patch(
    "/me/migration-board/posts/{post_id}",
    response_model=MyMigrationBoardPost,
)
def update_my_board_post(
    post_id: str,
    payload: UpdateMigrationBoardPostRequest,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    post = service.update_user_post(
        connection, current_user=current_user, post_id=post_id, payload=payload
    )
    connection.commit()
    return post


@router.post(
    "/me/migration-board/posts/{post_id}/submit",
    response_model=SubmitMigrationBoardPostResponse,
)
def submit_my_board_post(
    post_id: str,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    post = service.submit_user_post(
        connection, current_user=current_user, post_id=post_id
    )
    connection.commit()
    return {"post": post}


@router.post(
    "/me/migration-board/posts/{post_id}/archive",
    response_model=ArchiveMigrationBoardPostResponse,
)
def archive_my_board_post(
    post_id: str,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    post = service.archive_user_post(
        connection, current_user=current_user, post_id=post_id
    )
    connection.commit()
    return {"post": post}


@router.get(
    "/me/migration-board/matches",
    response_model=CompanionMatchesResponse,
)
def list_my_board_matches(
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
    limit: int = Query(20, ge=1, le=100),
) -> dict[str, Any]:
    return service.list_companion_matches(
        connection, current_user=current_user, post_id=None, limit=limit
    )


@router.get(
    "/me/migration-board/posts/{post_id}/matches",
    response_model=CompanionMatchesResponse,
)
def list_post_board_matches(
    post_id: str,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
    limit: int = Query(20, ge=1, le=100),
) -> dict[str, Any]:
    return service.list_companion_matches(
        connection, current_user=current_user, post_id=post_id, limit=limit
    )


@router.post(
    "/migration-board/posts/{post_id}/contact-requests",
    response_model=ContactRequestResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_board_contact_request(
    post_id: str,
    payload: CreateContactRequestRequest,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    request = service.create_contact_request(
        connection,
        current_user=current_user,
        post_id=post_id,
        message=payload.message,
    )
    connection.commit()
    return request


@router.get(
    "/me/migration-board/contact-requests/incoming",
    response_model=ContactRequestListResponse,
)
def list_incoming_contact_requests(
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    return service.list_incoming_requests(connection, current_user)


@router.get(
    "/me/migration-board/contact-requests/outgoing",
    response_model=ContactRequestListResponse,
)
def list_outgoing_contact_requests(
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    return service.list_outgoing_requests(connection, current_user)


@router.post(
    "/me/migration-board/contact-requests/{request_id}/accept",
    response_model=ContactRequestActionResponse,
)
def accept_contact_request(
    request_id: str,
    payload: ContactRequestActionRequest,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    request = service.accept_contact_request(
        connection,
        current_user=current_user,
        request_id=request_id,
        response_note=payload.response_note,
    )
    connection.commit()
    return {"request": request}


@router.post(
    "/me/migration-board/contact-requests/{request_id}/decline",
    response_model=ContactRequestActionResponse,
)
def decline_contact_request(
    request_id: str,
    payload: ContactRequestActionRequest,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    request = service.decline_contact_request(
        connection,
        current_user=current_user,
        request_id=request_id,
        response_note=payload.response_note,
    )
    connection.commit()
    return {"request": request}


@router.post(
    "/me/migration-board/contact-requests/{request_id}/cancel",
    response_model=ContactRequestActionResponse,
)
def cancel_contact_request(
    request_id: str,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    request = service.cancel_contact_request(
        connection, current_user=current_user, request_id=request_id
    )
    connection.commit()
    return {"request": request}


@router.post(
    "/migration-board/posts/{post_id}/report",
    response_model=MigrationBoardReportResponse,
    status_code=status.HTTP_201_CREATED,
)
def report_board_post(
    post_id: str,
    payload: CreateMigrationBoardReportRequest,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    report = service.report_post(
        connection,
        current_user=current_user,
        post_id=post_id,
        reason=payload.reason,
        details=payload.details,
    )
    connection.commit()
    return report


@router.post(
    "/me/migration-board/contact-requests/{request_id}/report",
    response_model=MigrationBoardReportResponse,
    status_code=status.HTTP_201_CREATED,
)
def report_board_contact_request(
    request_id: str,
    payload: CreateMigrationBoardReportRequest,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    report = service.report_contact_request(
        connection,
        current_user=current_user,
        request_id=request_id,
        reason=payload.reason,
        details=payload.details,
    )
    connection.commit()
    return report


@router.post(
    "/me/migration-board/blocks/{user_id}",
    response_model=BlockedUserResponse,
    status_code=status.HTTP_201_CREATED,
)
def block_board_user(
    user_id: str,
    payload: BlockUserRequest,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    row = service.block_user(
        connection,
        current_user=current_user,
        blocked_user_id=user_id,
        reason=payload.reason,
    )
    connection.commit()
    return row


@router.delete("/me/migration-board/blocks/{user_id}", status_code=204)
def unblock_board_user(
    user_id: str,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> None:
    service.unblock_user(
        connection, current_user=current_user, blocked_user_id=user_id
    )
    connection.commit()


@router.get(
    "/me/migration-board/blocks",
    response_model=BlockedUserListResponse,
)
def list_board_blocks(
    connection: Annotated[Connection[Any], Depends(get_connection)],
    current_user: Annotated[CurrentUser, Depends(require_user)],
) -> dict[str, Any]:
    return service.list_blocked_users(connection, current_user)
